oidc-api/src/main/java/uk/gov/di/authentication/oidc/services/InitiateIPVAuthorisationService.java

```diff
@@ -0,0 +1,149 @@
+import com.nimbusds.openid.connect.sdk.claims.ClaimsSetRequest;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import uk.gov.di.authentication.ipv.domain.IPVAuditableEvent;
+import uk.gov.di.authentication.ipv.entity.IPVAuthorisationResponse;
+import uk.gov.di.authentication.ipv.services.IPVAuthorisationService;
+import uk.gov.di.authentication.shared.entity.ClientRegistry;
+import uk.gov.di.authentication.shared.entity.ErrorResponse;
+import uk.gov.di.authentication.shared.entity.Session;
+import uk.gov.di.authentication.shared.entity.UserProfile;
+import uk.gov.di.authentication.shared.helpers.ClientSubjectHelper;
+import uk.gov.di.authentication.shared.helpers.IpAddressHelper;
+import uk.gov.di.authentication.shared.serialization.Json.JsonException;
+import uk.gov.di.authentication.shared.services.AuditService;
+import uk.gov.di.authentication.shared.services.AuthenticationService;
+import uk.gov.di.authentication.shared.services.CloudwatchMetricsService;
+import uk.gov.di.authentication.shared.services.ConfigurationService;
+
+import java.util.Map;
+import java.util.Optional;
+
+import static uk.gov.di.authentication.shared.helpers.ApiGatewayResponseHelper.generateApiGatewayProxyErrorResponse;
+import static uk.gov.di.authentication.shared.helpers.ApiGatewayResponseHelper.generateApiGatewayProxyResponse;
+import static uk.gov.di.authentication.shared.helpers.LogLineHelper.LogFieldName.CLIENT_ID;
+import static uk.gov.di.authentication.shared.helpers.LogLineHelper.attachLogFieldToLogs;
+import static uk.gov.di.authentication.shared.services.AuditService.MetadataPair.pair;
+
+public class InitiateIPVAuthorisationService {
+
+    private static final Logger LOG = LogManager.getLogger(InitiateIPVAuthorisationService.class);
+
+    private final ConfigurationService configurationService;
+    private final AuthenticationService authenticationService;
+    private final AuditService auditService;
+    private final IPVAuthorisationService authorisationService;
+    private final CloudwatchMetricsService cloudwatchMetricsService;
+
+    public InitiateIPVAuthorisationService(
+            ConfigurationService configurationService,
+            AuthenticationService authenticationService,
+            AuditService auditService,
+            IPVAuthorisationService authorisationService,
+            CloudwatchMetricsService cloudwatchMetricsService) {
+        this.configurationService = configurationService;
+        this.authenticationService = authenticationService;
+        this.auditService = auditService;
+        this.authorisationService = authorisationService;
+        this.cloudwatchMetricsService = cloudwatchMetricsService;
+    }
+
+    public APIGatewayProxyResponseEvent sendRequestToIPV(
+            APIGatewayProxyRequestEvent input,
+            AuthenticationRequest authRequest,
+            UserProfile userProfile,
+            Session session,
```

Comment:
Thoughts on how to reduce the number of parameters here would be to wrap some of them up into a class (or really just a record in java 17) containing all the user-y stuff from here, and then realised this [already exists](https://github.com/alphagov/di-authentication-api/blob/4df4c3c72d1c661dd481c287b3cb062a6d7a7ffe/shared/src/main/java/uk/gov/di/authentication/shared/state/UserContext.java#L12) and is used in the existing handler. Would that be appropriate to use here?

Reply:
Did think about using this as it was the way to bring them in before. However, was reluctant to use it as the authCallback lambda has already extracted out the information we need. We would need to manipulate our current data into `UserContext` and then when in our service do some manipulation to get it back to our original data. Example of this is `pairwiseSubject`.

Reply:
Fair enough :-)